lib/phoenix/sync/write.ex

```diff
@@ -0,0 +1,970 @@
+  import { createElectricSync } from '@TanStack/optimistic/electric';
+
+  // Create a collection configuration for todos
+  const todosConfig = {
+      id: 'todos',
+      // Create an ElectricSQL sync configuration
+      sync: createElectricSync(
+        {
+          // The `todo` shape is defined using `Phoenix.Sync.Router.sync/3` or `Phoenix.Sync.Controller.sync_render/3`
+          url: `http://localhost:4000/shapes/todo`,
+          params: {},
+        },
+        {
+          // Primary key for the todos table
+          primaryKey: ['id'],
+        }
+      ),
+    mutationFn: {
+      // Persist mutations to the backend
+      persist: async (mutations, transaction) => {
+        const response = await fetch(`http://localhost:4000/mutations`, {
+          method: `POST`,
+          headers: {
+            "content-type": `application/json`,
+            accept: `application/json`,
+          },
+          body: JSON.stringify({transaction: transaction.mutations}),
+        })
+        if (!response.ok) {
+          // Throwing an error will rollback the optimistic state.
+          throw new Error(`HTTP error! Status: ${response.status}`)
+        }
+
+        const result = await response.json()
+
+        return {
+          txid: result.txid,
+        }
+      },
+      // Wait for a transaction to be synced
+      awaitSync: async ({ config, persistResult: { txid: number } }) => {
+        try {
+          // Use the awaitTxid function from the ElectricSync configuration
+          // This waits for the specific transaction to be synced to the server
+          // The second parameter is an optional timeout in milliseconds
+          await config.sync.awaitTxid(persistResult.txid, 10000)
+          return true;
+        } catch (error) {
+          console.error('Error waiting for transaction to sync:', error);
+          // Throwing an error will rollback the optimistic state.
+          throw error;
+        }
+      },
+    },
+  };
+  ```
+
+  ## Server code
+
+  The `todos` table is represented by the `Ecto.Schema` module `Todo`.
+
+  ```elixir
+  defmodule Todo do
+    use Ecto.Schema
+
+    schema "todos" do
+      field :title, :string
+      field :completed, :boolean, default: false
+    end
+
+    def changeset(todo, data) do
+      todo
+      |> cast(data, [:id, :title, :completed])
+      |> validate_required([:id, :title])
+    end
+  end
+  ```
+
+  Our router maps the `/mutations` path to our `MutationsController.update/2`
+  function.
+
+  ```elixir
+  defmodule MutationsController do
+    use Phoenix.Controller, formats: [:json]
+
+    alias Phoenix.Sync.Write
+
+    def update(conn, %{transaction: transaction} = _params) do
+      {:ok, txid, _changes} =
+        Todo
+        |> Write.allow()
+        |> Write.apply(transaction)
+        |> Write.transaction(Repo)
+      render(conn, :update, txid: txid)
+    end
+  end
+  ```
+
+  - `allow/1` creates a write configuration that allows writes from the `todos`
```

Comment:
I don't really understand what's going on here. This allows function seems to be allowing all changes to the table. Is this some new authorization DSL? How does it integrate with per-transaction or per-mutation validation and authorization logic?

I see below:

> See `allow/3` for more information on the configuration options for each table.

Is that where more custom logic can be wired in?

Generally I think we have to be very careful about how and what we're introducing here and we must put it in context with the explanation first.